### ISSUE TITLE
Sealed configuration makes it much harder to configure Hikari

### ISSUE THREAD

**Issue by snicoll** [brettwooldridge/HikariCP#1095]
HikariCP version: `2.7.7`

The changes in #1053 in a patch release is quite surprising. Is that exception revealing a fundamentally broken behaviour or was it added to prevent a runtime change to the `DataSource` that was working (but somehow that feature had to be removed)?

Here is a use case: we are configuring Micrometer metrics in Spring Boot for each `HikariDataSource` found in the context. Metrics concerns are totally unrelated to `DataSource` configuration concern so, from a design perspective, they can't happen at the same location. I naively started to configure this and got caught by the fact the `DataSource` was sealed. The next step is to use a `BeanPostProcessor` to make sure that code is invoked before a `getConnection` call is invoked on the `DataSource`.

This is far from ideal and, again, that behaviour change in a patch release is very annoying. 

Besides, there is no way to know if a `DataSource` is sealed, so the only thing we can do is catch the exception and I am really not keen to do that. If configuring the `DataSource` is going to fail, we should be able to know about that.

**Comment by brettwooldridge** [comment]
@snicoll The change was made to enforce what has always been our policy; that the only supported method of changing pool configuration at runtime (i.e. once the pool as started) is through the ``HikariConfigMXBean`` -- and only the properties supported by that class.

We were getting too many support issues where users altered pool parameters (unsupported at runtime) only to find they had no effect.

However, the change in v2.7.7 also broke *special* behavior for the ``metricRegistry`` and ``metricsTrackerFactory`` settings.  We allow those to be set *one time*, whether the pool has started or not.  v2.7.7, by sealing the configuration upon ``HikariDataSource`` construction, broke this use-case.

That is now fixed in a commit I have on deck.  The sealed behavior regarding the rest of the pool configuration (unsupported properties) remains.  Methods defined by the ``HikariConfigMXBean`` interface remain unsealed.

I have also introduced two new methods, ``HikariDataSource.isRunning()`` and ``HikariConfig.isSealed()`` (which is also present on ``HikariDataSource`` by inheritence).

As per semantic versioning, I will release these changes as v3.0.0.

**Comment by brettwooldridge** [comment]
@snicoll I have published v3.0.0-SNAPSHOT to the sonatype snapshot repository.  Can you test it?

The repo maven settings are:
```xml
<repository>
   <id>oss-sonatype</id>
   <name>oss-sonatype</name>
   <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
   <snapshots>
      <enabled>true</enabled>
   </snapshots>
</repository>
```

**Comment by snicoll** [comment]
@brettwooldridge I will. Please do note that we are releasing Spring Boot 2.0 GA next week. If 3.0.0 is not available at that time we won't be able to upgrade until Spring Boot 3 (which would be, well, quite a disaster).

**Comment by wilkinsona** [comment]
Thanks, Brett. I've tried the 3.0.0 snapshots and it fixes the problem with the pool being sealed. In our case it was due to a call to `getConnection()` and such a call no longer prevents us from configuring the metrics tracker factory.

With regards to the need for a 3.0.0 for this change, I wonder if some other options could be considered? We have managed to work around the problems reported here and in #1096 with 2.7.7. If it's a choice between keeping those workarounds vs Hikari jumping to 3.0, I'd prefer to keep the workarounds. I also wondered if it would be possible to just fix the sealing problem that was introduced in 2.7.7 in a 2.7.8 release? That, too, would be preferable to a jump to 3.0 or even 2.8 just for this problem.
